tests/30rooms/04messages.pl

```diff
@@ -349,3 +349,55 @@
          })
       });
    };
+
+test "Ephemeral messages received from clients are correctly expired",
+   requires => [ local_user_and_room_fixtures() ],
+
+   do => sub {
+      my ( $user, $room_id ) = @_;
+
+      my $now_ms = int( time() * 1000 );
+      my $filter = '{"types":["m.room.message"]}';
+
+      matrix_send_room_message( $user, $room_id,
+         content => {
+            msgtype                          => "m.text",
+            body                             => "This is a message",
+            "org.matrix.self_destruct_after" => $now_ms + 1000,
+         },
+      )->then( sub {
+          matrix_get_room_messages( $user, $room_id, filter => $filter )
+      })->then( sub {
+         my ( $body ) = @_;
+         log_if_fail "Response body before expiry", $body;
+
+         my $chunk = $body->{chunk};
+         @$chunk == 1 or
+            die "Expected 1 message";
+
+         # Make sure we can read the message's content before it expires.
+         assert_eq( $chunk->[0]{content}{body}, "This is a message",
+            'chunk[0] content body' );
+
+         sleep( 2 );
+
+         retry_until_success {
```

Comment:
this really isn't great, because it means that the test will never fail properly, and worse, it won't tell you why it's failing, because the assertion error gets swallowed.

one way to improve it is to make sure that the error gets logged, with an `on_fail` handler:

```perl
my $iter = 0;
retry_until_success {
   $iter++;
   matrix_get_room_messages( ... ) ->then( sub {
      # ...
      log_if_fail "Iteration $iter: response body after expiry", $body;
      # ...
      Future->done(1);
   })->on_fail( sub {
      my ( $exc ) = @_;
      chomp $exc;
      log_if_fail "Iteration $iter: not ready yet: $exc";
   });
};